### ISSUE TITLE
Upgrade to systemd 238 changed predictable interface names (e.g. from wls1 to wlp2s0) for many Arch Linux users

### ISSUE THREAD

**Issue by dvorapa** [systemd/systemd#8446]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> 238

### Used distribution

> Arch Linux

### In case of bug report: Expected behaviour you didn't see

> …

### In case of bug report: Unexpected behaviour you saw

> Some years later my eth0 and wlan0 interfaces got renamed to wls1 and ens5 ad I fully understood that. But now the former predictable names changed to enp133s0 and wlp2s0, making them no more predictable at all. Which one of those, the former or the current is correct then?

> But more importantly it broke many networking tools, like wifi hotspot apps, internet connections, wifi connectors between devices and other stuff completely today. Not just for me, but also for the other on Arch Linux, see https://bbs.archlinux.org/viewtopic.php?id=235307.

### In case of bug report: Steps to reproduce the problem

> Upgrade from v237 to v238 on Arch Linux..

**Comment by floppym** [comment]
This could just as easily be caused by a kernel upgrade/downgrade. Have you isolated the issue to systemd?

**Comment by amishxda** [comment]
Before 238.0 i.e. with systemd 237.64

kernel is same but different name assigned by systemd

```
> uname -a
Linux amish 4.15.8-1-ARCH #1 SMP PREEMPT Sat Mar 10 00:00:33 UTC 2018 x86_64 GNU/Linux

> pacman -Q systemd
systemd 237.64-1

> journalctl -b 0 | grep renamed
Mar 14 10:21:49 amish kernel: sky2 0000:85:00.0 ens5: renamed from eth0
Mar 14 10:21:49 amish kernel: iwlwifi 0000:02:00.0 wls1: renamed from wlan0
```

After upgrade to 238.0 (no other package upgraded i.e. whole system otherwise is same)

```
> uname -a
Linux amish 4.15.8-1-ARCH #1 SMP PREEMPT Sat Mar 10 00:00:33 UTC 2018 x86_64 GNU/Linux

> pacman -Q systemd
systemd 238.0-3

> journalctl -b 0 | grep renamed
Mar 14 10:43:14 amish kernel: sky2 0000:85:00.0 enp133s0: renamed from eth0
Mar 14 10:43:14 amish kernel: iwlwifi 0000:02:00.0 wlp2s0: renamed from wlan0
```

Clearly predictable names changed!! i.e. they are no more predictable!

**Comment by amishxda** [comment]
0000:85:00.0 ---- 85 hex to decimal is 133 ... so it becomes p133 and s0 (enp133s0)
0000:02:00.0 ---- similarly this becomes wlp2s0

somehow systemd has started picking up these numbers for determining the name of certain network interfaces, which it didnt do earlier in my case and was picking up hot plug slot index number, instead (i.e. ens5 and wls1)

**Comment by amishxda** [comment]
Not sure but this commit might have caused the issue:
https://github.com/systemd/systemd/commit/73fc96c8ac0aa95477e53838a22ce94918c9d06f

298:  -                        if (strneq(address, udev_device_get_sysname(names->pcidev), strlen(address))) 
300: +                        if (streq(address, udev_device_get_sysname(names->pcidev))) 
299/301:                                  hotplug_slot = i;

May be now its failing and not setting hotplug_slot.

Or it could be that earlier it was a bug and now above commit unknowingly fixed that bug?

**Comment by lisaev** [comment]
Yeah, so much for the predictable interface names...

**Comment by filbranden** [comment]
This was clearly unintentional bug introduced as part of that refactor...

I pushed a fix with #8458, but I don't have the problem in my system, so if you can give that a try and confirm it fixes the issue for you that would be great!

**Comment by amishxda** [comment]
I am not sure what "hotplug" here refers to.

I use laptop and my interfaces are internal. And I can not hotplug (or remove) them at my will. (but then technology used may be hotplug based which I am unaware of i.e. if my laptop was open i may actually cud hotplug or remove the interface)

So depending on what actually is hotplug for systemd. The actual commit (made in February) may actually have fixed an old unknown bug, unintentionally.

If not then it is indeed bug and may need reverting.
